feat: Coverage converter publishing/usage document (#713)
platform-platform/monorepo: metrics/coverage_converter/docs/02_coverage_converter_publishing.md

2. Create a new release.
3. Add executables to created release.

Moreover, GitHub Actions has ready-to-use actions that simplify the process of publishing: 

- [Create a release action](https://github.com/marketplace/actions/create-a-release) - an action that helps to create a new release.
- [Upload a release asset action](https://github.com/marketplace/actions/upload-a-release-asset) - an action that helps to add assets to the existing release.

### Pros & Cons

Pros: 
- Easy to configure.
- Great integration with GitHub Actions.
- Does not require any configuration files to publish.
- Can publish any file type, so downloading the CLI user gets the ready-to-use application.

Cons: 
- More complex configuration for monorepo (could be solved with tags, etc..)
- Downloading to CI action becomes a bit more complex compared to the [alternatives](#Personal-Package-Archive-(PPA)-Homebrew-and-Chocolatey), see more [here](#Downloading-using-the-command-line-commands)

So, the GitHub Releases is the best choice in our case because it allows us easily configure the publishing and allows users to easily download and start using the Coverage Converter tool. 

### Publishing details

> Once we've chosen the [GitHub Releases](#GitHub-Releases), let's consider the deployment process in more detail. To publish the application to the GitHub releases, we should create a tag that will correspond to the release. Since we want to keep the release history and we'll have more than one project on our repository, we should create a tag that will contain the version number and the name of the project - `v1.0-coverage-converter`. Also, we want to publish the `pre-release` versions, so we should create a tag for those releases. Since we don't want to keep the `pre-release` versions history, we can create a `coverage-converter-snapshot` tag that will be re-assigned on each release.

Reviewer: ```suggestion
Once we've chosen the [GitHub Releases](#GitHub-Releases), let's consider the deployment process in more detail. To publish the application to the GitHub releases, we should create a tag that will correspond to the release. Since we want to keep the release history and we'll have more than one project on our repository, we should create a tag that will contain the version number and the name of the project - `v1.0-coverage-converter`.

For every time that we change coverage converter tool sources, we use a `coverage-converter-snapshot` tag that will be re-assigned after each build to the newly published binary. It should be marked as `pre-release`. There should be only 1 release at a time with `coverage-converter-snapshot` tag.
```
Would that make sense?

# Coverage Converter usage

First of all, you should download the Coverage Converter tool. There are a couple of ways to do that:

- Download using the GitHub GUI.
- Download using the command line commands.

Let's consider each of them separately: 

> ## Downloading using the GitHub GUI

Reviewer: Would it make sense to mention downloading the latest snapshot here? As that would be much easier - just use the URL that we know now.

To download the Coverage Converter tool from the GitHub releases, follow the next steps: 

1. Go to the platform-platform/monorepo GitHub repository [releases page](https://github.com/platform-platform/monorepo/releases). 
2. Find the version of the Coverage Converter tool you want to download.
3. Open the `Assets` collapsible section and choose the Coverage Converter file compatible with your OS.

> 4. Click to download

Reviewer: ```suggestion
4. Click to download.
```

## Downloading using the command line commands

If you want to download the Coverage Converter tool you should follow the steps below: 

1. Open the GitHub repository releases and find the release you want to download, as described in the [Downloading using the GitHub GUI](#Downloading-using-the-GitHub-GUI) section.

> 2. Find the executable file compatible with your OS

Reviewer: ```suggestion
2. Find the executable file compatible with your OS.
```

- `PPA` for the Linux users
- `Homebrew` for the macOS users
- `Chocolatey` for Windows users

This alternative is pretty complex in configuration, but it provides an easy way of installing the Coverage Converter tool using the single platform-unified command. So, let's consider the main Pros & Cons of this alternative: 

### Pros & Cons

Pros: 
- Easy to download the latest release.
- Makes the CLI global-accessible command after installation.

Cons: 
- Complex configuration process that requires a full configuration process for each platform (package manager).
- Some package management tools do not support publishing the compiled executables, so we should compile it on the user's machine.
- No existing integration with GitHub Actions.

## pub.dev publishing

Another alternative is to publish the Coverage Converter tool as a Dart package. It will allow users to install the CLI tool using the `pub global activate` command, but it requires the Dart SDK installed.

To publish the Dart package, you should: 

1. Create a Dart package project.
2. Publish the package using the `pub publish` command.

Let's consider the pros and cons of this publishing approach: 

### Pros & Cons

Pros: 

> - Easy to publish 
> - Easy to download and install the command-line application
> - Single publishing for all platforms
> - Does not require building before publishing

Reviewer: ```suggestion
- Easy to publish.
- Easy to download and install the command-line application.
- Single publishing for all platforms.
- Does not require building before publishing.
```

# Coverage Converter publishing

> Summary of the proposed change

An explanation of the Coverage Converter tool publishing place and process

# References

> Link to supporting documentation, GitHub tickets, etc.

- [GitHub releases](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/managing-releases-in-a-repository)

> - [PPA](https://help.ubuntu.com/stable/ubuntu-help/addremove-ppa.html.en#:~:text=Personal%20Package%20Archives%20(PPAs)%20are,that%20it%20can%20be%20tested.)

Reviewer: ```suggestion
- [PPA](https://help.ubuntu.com/stable/ubuntu-help/addremove-ppa.html.en)
```

The GitHub provides an easy way of publishing the applications - GitHub Releases. It allows you to create a GitHub release and publish any files you want there. So, to publish a Coverage Converter CLI, we'll need to follow the next steps: 

1. Build the Coverage Converter tool.
2. Create a new release.
3. Add executables to created release.

Moreover, GitHub Actions has ready-to-use actions that simplify the process of publishing: 

- [Create a release action](https://github.com/marketplace/actions/create-a-release) - an action that helps to create a new release.
- [Upload a release asset action](https://github.com/marketplace/actions/upload-a-release-asset) - an action that helps to add assets to the existing release.

### Pros & Cons

Pros: 
- Easy to configure.
- Great integration with GitHub Actions.
- Does not require any configuration files to publish.
- Can publish any file type, so downloading the CLI user gets the ready-to-use application.

Cons: 
- Requires a separate repository for each project. 
- Should be built on all supported platforms to publish the binary.
- Downloading on the CI can be a bit complex because it requires manual downloading and installation.

> So, the GitHub Releases is the best choice in our case because it allows us easily configure the publishing mechanism and allows users to comparatively easy download and start using the Coverage Converter tool. 

Reviewer: ```suggestion
So, the GitHub Releases is the best choice in our case because it allows us easily configure the publishing and allows users to easily download and start using the Coverage Converter tool. 
```

> # Coverage Converter usage

Reviewer: Can we add a little bit more details about publishing - like what tags we need to use in releases when using monorepo (our case) or how basically we can distinguish between different CLI tools?

3. Open the `Assets` collapsible section and choose the Coverage Converter file compatible with your OS.
4. Click to download

## Downloading using the command line commands

If you want to download the Coverage Converter tool from the CI, for examples, or just using the command line, you should follow the steps below: 

1. Open the GitHub repository releases and find the release you want to download, as described in the [Downloading using the GitHub GUI](#Downloading-using-the-GitHub-GUI) section.
2. Find the executable file compatible with your OS
3. Right-click on this file and copy the link address. 

After these steps, you should obtain the download link similar to this one: 

`https://github.com/platform-platform/monorepo/releases/download/v1.0.0/coverage_converter_macos`

As you can see, the link contains the release version. You can modify the link to make it refer to the latest release by removing the version from the URL and adding the `latest` keyword after the `releases`. So, the link should look like this: 

`https://github.com/platform-platform/monorepo/releases/latest/download/coverage_converter_macos`

## Using the Coverage Converter tool

> Once you've obtained the downloading link, you can download the Coverage Converter tool using the command line commands. Let's consider the macOS command-line command to download the Coverage Converter tool: 

Reviewer: ```suggestion
Once you've obtained the download link, you can download the Coverage Converter tool using the CLI tools. Let's consider the macOS command to download the Coverage Converter tool: 
```

`curl -o <output> -k <url>`

Where the `<output>` is the file path where you want to download the coverage converter, and the `<url>` is the download URL obtained previously. 

So, when you've got the Coverage Converter executable, you probably, want to convert coverage to be readable by the CI integrations tool. To run the conversion process, you should have the coverage report in one of the supported formats: 

- `LCOV`
- `Istanbul`

Let's consider the command for converting the `LCOV` format for macOS: 

`./coverage_converter_macos lcov -i <YOUR_COVERAGE_REPORT_FILE> -o <COVERAGE_REPORT_OUTPUT_FILE.json>`

As you can see, the Coverage Converter CLI contains a separate command for each supported coverage format. So, to convert the `Istanbul` coverage report, you should replace the `lcov` command with `istanbul` and so on. Also, let's review the CLI arguments: 

- `-i, --input` - the parameter to specify the input coverage file path. This parameter is required, and the CLI cannot be run without this argument.

> - `-o, --output` - the parameter to specify the output coverage file path. This parameter could be 

Reviewer: ```suggestion
- `-o, --output` - the parameter to specify the output coverage file path. This parameter could be omitted.
```

Maybe we should link to the Coverage Converter tool design here so we don't duplicate parameter description?

After the conversion process is finished, the command will create a `COVERAGE_REPORT_OUTPUT_FILE.json` file with the coverage report JSON readable by the CI integrations tool.

# Dependencies

> What is the project blocked on?

This project has no blockers.

> What will be impacted by the project?

Impacts the publishing approach of the Coverage Converter tool.

# Testing

> How will the project be tested?

This project will be tested manually. 

# Alternatives Considered

> Summarize alternative designs (pros & cons)

# Coverage Converter publishing

## Personal Package Archive (PPA), Homebrew, and Chocolatey.

The alternative of the GitHub Releases is to publish the CLI to the package managers of different platforms. So, we'll deploy the CLI to the following package managers: 

- `PPA` for the Linux users
- `Homebrew` for the macOS users
- `Chocolatey` for Windows users

> This alternative is pretty complex in configuration, but it provides an easy way of installing the CLI using the single platform-unified command. So, let's consider the main Pros & Cons of this alternative: 

Reviewer: ```suggestion
This alternative is pretty complex in configuration, but it provides an easy way of installing the Coverage Converter tool using the single platform-unified command. So, let's consider the main Pros & Cons of this alternative: 
```

### Pros & Cons

Pros: 
- Easy to download the latest release.
- Makes the CLI global-accessible command after installation.

Cons: 
- Complex configuration process.
- Requires a separate configuration file for each platform (package management tool).
- Some package management tools do not support publishing the compiled executables, so we should compile it on the user's machine.
- Bad integration with GitHub Actions.

## pub.dev publishing

> Another alternative is to publish the Coverage Converter tool as a Dart package. It will allow users to install the CLI using the `pub global activate` command, but it requires the Dart SDK installed. 

Reviewer: ```suggestion
Another alternative is to publish the Coverage Converter tool as a Dart package. It will allow users to install the CLI tool using the `pub global activate` command, but it requires the Dart SDK installed. 
```

- `LCOV`
- `Istanbul`

Let's consider the command for converting the `LCOV` format for macOS: 

`./coverage_converter_macos lcov -i <YOUR_COVERAGE_REPORT_FILE> -o <COVERAGE_REPORT_OUTPUT_FILE.json>`

As you can see, the Coverage Converter CLI contains a separate command for each supported coverage format. So, to convert the `Istanbul` coverage report, you should replace the `lcov` command with `istanbul` and so on. See [Coverage Converter design](https://github.com/platform-platform/monorepo/blob/master/metrics/coverage_converter/docs/01_coverage_converter_design.md#cli-design) document for detailed description of the CLI arguments.

After the conversion process is finished, the command will create a `COVERAGE_REPORT_OUTPUT_FILE.json` file with the coverage report JSON readable by the CI integrations tool.

# Dependencies

> What is the project blocked on?

This project has no blockers.

> What will be impacted by the project?

Impacts the publishing approach of the Coverage Converter tool.

# Testing

> How will the project be tested?

This project will be tested manually. 

# Alternatives Considered

> Summarize alternative designs (pros & cons)

> # Coverage Converter publishing

Reviewer: ```suggestion
```
As the document is about publishing it means that we can skip it here?

## Personal Package Archive (PPA), Homebrew, and Chocolatey.

The alternative of the GitHub Releases is to publish the CLI to the package managers of different platforms. So, we'll deploy the CLI to the following package managers: 

- `PPA` for the Linux users
- `Homebrew` for the macOS users
- `Chocolatey` for Windows users

This alternative is pretty complex in configuration, but it provides an easy way of installing the Coverage Converter tool using the single platform-unified command. So, let's consider the main Pros & Cons of this alternative: 

### Pros & Cons

Pros: 
- Easy to download the latest release.
- Makes the CLI global-accessible command after installation.

Cons: 

> - Complex configuration process.

Reviewer: ```suggestion
- Complex configuration process: no default Github actions.
```
Or maybe there are? Or did you mean something else here?

> - Requires a separate configuration file for each platform (package management tool).

Reviewer: ```suggestion
- Requires a full configuration process for each platform (package manager).
```

- Some package management tools do not support publishing the compiled executables, so we should compile it on the user's machine.

> - Bad integration with GitHub Actions.

Reviewer: ```suggestion
- No existing integration with GitHub Actions.
```
Or can we be a bit more specific about what "bad" means here given your understanding of the process?
